### ISSUE TITLE
Undefined references in the runtime library

### ISSUE THREAD

**Issue by invexed** [AdaptiveCpp/AdaptiveCpp#589]
My program compiles successfully but I'm getting lots of linker errors. For example:

```cpp
[100%] Linking CXX executable my-exe
/usr/sbin/ld: CMakeFiles/my-exe.dir/src/main.cpp.o: in function `hipsycl::sycl::detail::buffer_impl::~buffer_impl()':
main.cpp:(.text._ZN7hipsycl4sycl6detail11buffer_implD2Ev[_ZN7hipsycl4sycl6detail11buffer_implD2Ev]+0x361): undefined reference to `hipsycl::rt::dag_builder::add_explicit_mem_requirement(std::__1::unique_ptr<hipsycl::rt::operation, std::__1::default_delete<hipsycl::rt::operation> >, hipsycl::rt::requirements_list const&, hipsycl::rt::execution_hints const&)'
```
All of the undefined references seem to be in ```hipSYCL-rt```:
```cpp
undefined reference to `hipsycl::rt::requirements_list::add_requirement(std::__1::unique_ptr<hipsycl::rt::requirement, std::__1::default_delete<hipsycl::rt::requirement> >)'
undefined reference to `hipsycl::rt::memory_location::memory_location(hipsycl::rt::device_id, hipsycl::rt::static_array<3>, std::__1::shared_ptr<hipsycl::rt::data_region<void*> >)'
undefined reference to `hipsycl::rt::execution_hints::add_hint(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
...
```

I've tried explicitly linking ```hipSYCL-rt``` in ```target_link_libraries``` but to no avail. Any idea what could be causing them?

```Dockerfile```:
```dockerfile
FROM archlinux:base

RUN pacman -Syu --noconfirm cmake gcc libc++ make wget \
&& wget -O - https://raw.githubusercontent.com/illuhad/hipSYCL/develop/install/scripts/add-hipsycl-repo/archlinux-rolling.sh | sh \
&& pacman -S --noconfirm hipSYCL-base-nightly hipSYCL-omp-nightly hipSYCL-cuda-nightly \
&& wget -O - https://raw.githubusercontent.com/illuhad/hipSYCL/develop/install/scripts/install-cuda.sh | sh \
&& pacman -Scc --noconfirm

RUN groupadd -r name && useradd --no-log-init -r -g name name
USER name:name
```

```CMakeLists.txt```:
```cmake
cmake_minimum_required(VERSION 3.15)
project(my-proj CXX)

find_package(hipSYCL REQUIRED CONFIG)

add_executable(my-exe src/main.cpp)

target_include_directories(my-exe PRIVATE include)
target_link_libraries(my-exe PRIVATE -stdlib=libc++)
target_compile_definitions(my-exe PRIVATE
    HIPSYCL_DEBUG_LEVEL=${HIPSYCL_DEBUG_LEVEL}
    HIPSYCL_STRICT_ACCESSOR_DEDUCTION)
target_compile_options(my-exe PRIVATE
    -fcuda-approx-transcendentals
    -fcuda-flush-denormals-to-zero
    -ffp-contract=fast
    -stdlib=libc++)

add_sycl_to_target(TARGET my-exe)
```
I'm invoking CMake like so:
```
cmake -S . -B build -D CMAKE_BUILD_TYPE=Release -D CMAKE_CXX_COMPILER=/opt/hipSYCL/llvm/llvm/bin/clang++ -D hipSYCL_DIR=/opt/hipSYCL/lib/cmake/hipSYCL -D HIPSYCL_DEBUG_LEVEL=2 -D HIPSYCL_TARGETS=cuda:sm_50 && cmake --build build --target my-exe
```

**Comment by sbalint98** [comment]
I think this error is due to the fact that the archlinux packages are outdated. They were built against an older version of the c++ standard library. This is related to the issue of building llvm11 with gcc11, that I have mentioned here: #586  

> && pacman -S --noconfirm hipSYCL-base-nightly hipSYCL-omp-nightly hipSYCL-cuda-nightly \

`hipSYCL-base-nightly` is from before hipSYCL has been switched to the current modular backend system, and is very old. Please install `hipSYCL-cuda-nightly` only. It will pull in all the necessary dependencies, eg: `hipSYCL-core-nightly`, `hipSYCL-base-nightly` and `hipSYCL-omp-nightly`. After the installation you will still need to execute the `/install/scripts/isntall-cuda.sh` in order to install the Cuda-toolkit, since `hipSYCL-cuda-nightly` only contains the hipSYCL backend library.

**Comment by illuhad** [comment]
@sbalint98 Could it not also be related to the fact that @invexed is trying to link the application against `libc++` while `hipSYCL-rt` probably was linked against `libstdc++`? Since the `hipSYCL-rt` has a C++ ABI that relies on standard types, I would imagine that using a different standard library than the one hipSYCL was built against might break ABI.

**Comment by invexed** [comment]
I've modified my ```Dockerfile``` to only install ```hipSYCL-cuda-nightly``` but the linker errors are still there. I think @illuhad is right about ```libc++``` not being ABI-compatible with ```libstdc++``` since all of the undefined references are functions that either accept at least one type from ```std::``` or that have no parameters (I presume it is the return type in these cases that is from ```std::```).

To be honest I would prefer to use a recent ```libstdc++``` and that is what I tried first. However the ```libstdc++``` that comes with GCC 11 is not usable with the Clang 11.0.1 in the ```hipSYCL``` repos (the errors are different to the one mentioned in [spack/spack#24435](https://github.com/spack/spack/issues/24435); e.g. Clang hasn't implemented the "down with typename" proposal required to compile the ```<ranges>``` header, which is being included by some other standard library header that I _am_ including).

I also remember having trouble using the ```libstdc++``` that comes with GCC 10 when I first updated to Clang 10, but it might work with 11. I'll give it a try and report back.
